modules/apps/29-fee/keeper/msg_server.go

```diff
@@ -0,0 +1,67 @@
+package keeper
+
+import (
+	"context"
+
+	sdk "github.com/cosmos/cosmos-sdk/types"
+
+	"github.com/cosmos/ibc-go/v3/modules/apps/29-fee/types"
+	channeltypes "github.com/cosmos/ibc-go/v3/modules/core/04-channel/types"
+)
+
+var _ types.MsgServer = Keeper{}
+
+// RegisterCounterpartyAddress is called by the relayer on each channelEnd and allows them to specify their counterparty address before relaying
+// This ensures they will be properly compensated for forward relaying on the source chain since the destination chain must send back relayer's source address (counterparty address) in acknowledgement
+// This function may be called more than once by relayers, in which case, the previous counterparty address will be overwritten by the new counterparty address
+func (k Keeper) RegisterCounterpartyAddress(goCtx context.Context, msg *types.MsgRegisterCounterpartyAddress) (*types.MsgRegisterCounterpartyAddressResponse, error) {
+	ctx := sdk.UnwrapSDKContext(goCtx)
+
+	k.SetCounterpartyAddress(
```

Comment:
> We may want to validate that msg.Address is a valid address for the local chain.

This will already be checked during antehandler, since address must sign message.

Do you have suggestions on Counterparty address validation?